File: src/NodeParser/MappedTypeNodeParser.ts
import ts from "typescript";
import { ExpectationFailedError } from "../Error/Errors.js";
import type { NodeParser } from "../NodeParser.js";
import { Context } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import { AnnotatedType } from "../Type/AnnotatedType.js";
import { AnyType } from "../Type/AnyType.js";
import { ArrayType } from "../Type/ArrayType.js";
import type { BaseType } from "../Type/BaseType.js";
import { DefinitionType } from "../Type/DefinitionType.js";
import type { EnumValue } from "../Type/EnumType.js";
import { EnumType } from "../Type/EnumType.js";
import { LiteralType } from "../Type/LiteralType.js";
import { NeverType } from "../Type/NeverType.js";
import { NumberType } from "../Type/NumberType.js";
import { ObjectProperty, ObjectType } from "../Type/ObjectType.js";
import { StringType } from "../Type/StringType.js";
import { SymbolType } from "../Type/SymbolType.js";
import { TupleType } from "../Type/TupleType.js";
import { UnionType } from "../Type/UnionType.js";
import { derefAnnotatedType, derefType, isDeepLiteralUnion } from "../Utils/derefType.js";
import { getKey } from "../Utils/nodeKey.js";
import { preserveAnnotation } from "../Utils/preserveAnnotation.js";
import { removeUndefined } from "../Utils/removeUndefined.js";
import { uniqueTypeArray } from "../Utils/uniqueTypeArray.js";

export class MappedTypeNodeParser implements SubNodeParser {
    public constructor(
        protected childNodeParser: NodeParser,
        protected readonly additionalProperties: boolean,
    ) {}

    public supportsNode(node: ts.MappedTypeNode): boolean {
        return node.kind === ts.SyntaxKind.MappedType;
    }

    public createType(node: ts.MappedTypeNode, context: Context): BaseType {
        const constraintType = this.childNodeParser.createType(node.typeParameter.constraint!, context);
        const keyListType = derefType(constraintType);
        const id = `indexed-type-${getKey(node, context)}`;
        const isKeyOf = this.isKeyOfConstraint(node);

        if (keyListType instanceof UnionType) {
            // When the constraint is `keyof T` where T is a tuple, the keys are sequential
            // numeric literals (0, 1, 2, ...). Preserve the tuple structure.
            if (isKeyOf) {
                const tupleIndices = this.getNumericLiteralIndices(keyListType);
                if (tupleIndices) {
                    const items = tupleIndices.map((idx) =>
                        this.childNodeParser.createType(node.type!, this.createSubContext(node, idx, context)),
                    );
                    return new TupleType(items);
                }
            }

            // Key type resolves to a set of known properties
            return new ObjectType(
                id,
                [],
                this.getProperties(node, keyListType, context),
                this.getAdditionalProperties(node, keyListType, context),
            );
        }

        if (keyListType instanceof LiteralType) {
            // Key type resolves to single known property
            return new ObjectType(id, [], this.getProperties(node, new UnionType([keyListType]), context), false);
        }

        const maybeUnionType = this.childNodeParser.createType(
            node.type!,
            this.createSubContext(node, keyListType, context),
        );
        if (constraintType?.getId() === "number" && !(maybeUnionType instanceof NeverType)) {
            // When keyof resolves to number (array) or the value is a union, produce an array.
            if (maybeUnionType instanceof UnionType || isKeyOf) {
                return new ArrayType(maybeUnionType);
            }
        }

        if (
            keyListType instanceof StringType ||
            keyListType instanceof NumberType ||
            keyListType instanceof SymbolType ||
            keyListType instanceof AnyType
        ) {
            // Key type widens to `string`
            const type = this.childNodeParser.createType(node.type!, this.createSubContext(node, keyListType, context));
            const resultType = new ObjectType(id, [], [], type);
            let annotations;
            if (constraintType instanceof AnnotatedType) {
                annotations = constraintType.getAnnotations();
            } else if (constraintType instanceof DefinitionType) {
                const childType = constraintType.getType();
                if (childType instanceof AnnotatedType) {
                    annotations = childType.getAnnotations();
                }
            }
            if (annotations) {
                return new AnnotatedType(resultType, { propertyNames: annotations }, false);
            }
            return resultType;
        }

        if (keyListType instanceof EnumType) {
            return new ObjectType(id, [], this.getValues(node, keyListType, context), false);
        }

        if (keyListType instanceof NeverType) {
            return new ObjectType(id, [], [], false);
        }

        throw new ExpectationFailedError(
            `Unexpected key type "${
                constraintType ? constraintType.getId() : constraintType
            }" for this node. (expected "UnionType" or "StringType")`,
            node,
        );
    }

    /**
     * In mapped types, questionToken can be:
     * - undefined: no optional modifier
     * - QuestionToken (?): add optional
     * - PlusToken (+?): add optional
     * - MinusToken (-?): remove optional (e.g. Required<T>) → property is required
     */
    protected isMappedPropertyRequired(node: ts.MappedTypeNode, hasUndefinedInType: boolean): boolean {
        if (node.questionToken === undefined) {
            return !hasUndefinedInType;
        }
        if (node.questionToken.kind === ts.SyntaxKind.MinusToken) {
            return true; // -? removes optional → output property is always required
        }
        return false;
    }

    protected mapKey(node: ts.MappedTypeNode, rawKey: LiteralType, context: Context): BaseType {
        if (!node.nameType) {
            return rawKey;
        }
        return derefType(this.childNodeParser.createType(node.nameType, this.createSubContext(node, rawKey, context)));
    }

    protected getProperties(node: ts.MappedTypeNode, keyListType: UnionType, context: Context): ObjectProperty[] {
        return uniqueTypeArray(keyListType.getFlattenedTypes(derefType))
            .filter((type): type is LiteralType => type instanceof LiteralType)
            .map((type) => [type, this.mapKey(node, type, context)])
            .filter((value): value is [LiteralType, LiteralType] => value[1] instanceof LiteralType)
            .reduce((result: ObjectProperty[], [key, mappedKey]: [LiteralType, LiteralType]) => {
                const propertyType = this.childNodeParser.createType(
                    node.type!,
                    this.createSubContext(node, key, context),
                );

                let newType = derefAnnotatedType(propertyType);
                let hasUndefined = false;
                if (newType instanceof UnionType) {
                    const { newType: newType_, numRemoved } = removeUndefined(newType);
                    hasUndefined = numRemoved > 0;
                    newType = newType_;
                }

                const objectProperty = new ObjectProperty(
                    mappedKey.getValue().toString(),
                    preserveAnnotation(propertyType, newType),
                    this.isMappedPropertyRequired(node, hasUndefined),
                );

                result.push(objectProperty);
                return result;
            }, []);
    }

    protected getValues(node: ts.MappedTypeNode, keyListType: EnumType, context: Context): ObjectProperty[] {
        return keyListType
            .getValues()
            .filter((value: EnumValue) => value != null)
            .map((value: EnumValue) => {
                const type = this.childNodeParser.createType(
                    node.type!,
                    this.createSubContext(node, new LiteralType(value!), context),
                );

                return new ObjectProperty(value!.toString(), type, this.isMappedPropertyRequired(node, false));
            });
    }

    protected getAdditionalProperties(
        node: ts.MappedTypeNode,
        keyListType: UnionType,
        context: Context,
    ): BaseType | boolean {
        if (isDeepLiteralUnion(keyListType)) {
            return this.additionalProperties;
        }

        const key = keyListType.getTypes().filter((type) => !(derefType(type) instanceof LiteralType))[0];

        if (key) {
            return (
                this.childNodeParser.createType(node.type!, this.createSubContext(node, key, context)) ??
                this.additionalProperties
            );
        }

        return this.additionalProperties;
    }

    protected isKeyOfConstraint(node: ts.MappedTypeNode): boolean {
        const c = node.typeParameter.constraint;
        return (
            c?.kind === ts.SyntaxKind.TypeOperator &&
            (c as ts.TypeOperatorNode).operator === ts.SyntaxKind.KeyOfKeyword
        );
    }

    /** Returns sorted numeric literal indices if all keys are numeric (tuple keys), else undefined. */
    protected getNumericLiteralIndices(keyListType: UnionType): LiteralType[] | undefined {
        const types = keyListType.getFlattenedTypes(derefType);
        if (types.length === 0) return undefined;
        const indices: LiteralType[] = [];
        for (const t of types) {
            if (!(t instanceof LiteralType) || typeof t.getValue() !== "number") return undefined;
            indices.push(t);
        }
        return indices.sort((a, b) => (a.getValue() as number) - (b.getValue() as number));
    }

    protected createSubContext(
        node: ts.MappedTypeNode,
        key: LiteralType | StringType | NumberType,
        parentContext: Context,
    ): Context {
        const subContext = new Context(node);

        for (const parentParameter of parentContext.getParameters()) {
            subContext.pushParameter(parentParameter);
            subContext.pushArgument(parentContext.getArgument(parentParameter));
        }

        subContext.pushParameter(node.typeParameter.name.text);
        subContext.pushArgument(key);

        return subContext;
    }
}

File: internal/parser/mapped_type_node_parser.go
package parser

// Port of src/NodeParser/MappedTypeNodeParser.ts.

import (
	"fmt"

	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

type MappedTypeNodeParser struct {
	childNodeParser      NodeParser
	additionalProperties bool
}

func NewMappedTypeNodeParser(childNodeParser NodeParser, additionalProperties bool) *MappedTypeNodeParser {
	return &MappedTypeNodeParser{
		childNodeParser:      childNodeParser,
		additionalProperties: additionalProperties,
	}
}

func (p *MappedTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindMappedType
}

func (p *MappedTypeNodeParser) CreateType(node *ast.Node, context *Context, _ *types.ReferenceType) types.Type {
	mapped := node.AsMappedTypeNode()
	constraintType := p.childNodeParser.CreateType(mapped.TypeParameter.AsTypeParameterDeclaration().Constraint, context, nil)
	keyListType := types.DerefType(constraintType)
	id := "indexed-type-" + GetNodeKey(node, context)

	if keyListUnion, ok := keyListType.(*types.UnionType); ok {
		// Key type resolves to a set of known properties.
		return types.NewObjectType(
			id,
			nil,
			p.getProperties(node, keyListUnion, context),
			p.getAdditionalProperties(node, keyListUnion, context),
			false,
		)
	}

	if keyListLiteral, ok := keyListType.(*types.LiteralType); ok {
		// Key type resolves to a single known property.
		return types.NewObjectType(id, nil, p.getProperties(node, types.NewUnionType([]types.Type{keyListLiteral}), context), false, false)
	}

	maybeUnionType := p.childNodeParser.CreateType(mapped.Type, p.createSubContext(node, keyListType, context), nil)
	if maybeUnion, ok := maybeUnionType.(*types.UnionType); ok && constraintType != nil && constraintType.ID() == "number" {
		// Then we turn it into an array.
		return &types.ArrayType{Item: maybeUnion}
	}

	switch keyList := keyListType.(type) {
	case *types.StringType, *types.NumberType, *types.SymbolType, *types.AnyType:
		// Key type widens to `string`.
		valueType := p.childNodeParser.CreateType(mapped.Type, p.createSubContext(node, keyListType, context), nil)
		resultType := types.NewObjectType(id, nil, nil, valueType, false)

		var annotations types.Annotations
		if annotated, ok := constraintType.(*types.AnnotatedType); ok {
			annotations = annotated.Annotations
		} else if definition, ok := constraintType.(*types.DefinitionType); ok {
			if annotated, ok := definition.Type.(*types.AnnotatedType); ok {
				annotations = annotated.Annotations
			}
		}
		if annotations != nil {
			return &types.AnnotatedType{
				Type:        resultType,
				Annotations: types.Annotations{"propertyNames": annotations},
				Nullable:    false,
			}
		}
		return resultType

	case *types.EnumType:
		return types.NewObjectType(id, nil, p.getValues(node, keyList, context), false, false)

	case *types.NeverType, *types.HiddenType:
		// HiddenType subclasses NeverType upstream, so it matches this
		// branch there as well.
		return types.NewObjectType(id, nil, nil, false, false)
	}

	constraintID := "undefined"
	if constraintType != nil {
		constraintID = constraintType.ID()
	}
	panic(fmt.Errorf("unexpected key type %q for this node. (expected \"UnionType\" or \"StringType\") %s", constraintID, DescribeNode(node)))
}

// isMappedPropertyRequired evaluates the optional modifier of the mapped
// type. In mapped types, questionToken can be:
//   - absent: no optional modifier
//   - `?` (QuestionToken): add optional
//   - `+?` (PlusToken): add optional
//   - `-?` (MinusToken): remove optional (e.g. Required<T>) -> required
func (p *MappedTypeNodeParser) isMappedPropertyRequired(node *ast.Node, hasUndefinedInType bool) bool {
	questionToken := node.AsMappedTypeNode().QuestionToken
	if questionToken == nil {
		return !hasUndefinedInType
	}
	if questionToken.Kind == ast.KindMinusToken {
		return true // -? removes optional -> output property is always required
	}
	return false
}

func (p *MappedTypeNodeParser) mapKey(node *ast.Node, rawKey *types.LiteralType, context *Context) types.Type {
	nameType := node.AsMappedTypeNode().NameType
	if nameType == nil {
		return rawKey
	}
	return types.DerefType(p.childNodeParser.CreateType(nameType, p.createSubContext(node, rawKey, context), nil))
}

func (p *MappedTypeNodeParser) getProperties(node *ast.Node, keyListType *types.UnionType, context *Context) []*types.ObjectProperty {
	mapped := node.AsMappedTypeNode()
	var result []*types.ObjectProperty

	for _, keyType := range types.UniqueTypes(keyListType.FlattenedTypes(types.DerefType)) {
		key, ok := keyType.(*types.LiteralType)
		if !ok {
			continue
		}
		mappedKey, ok := p.mapKey(node, key, context).(*types.LiteralType)
		if !ok {
			continue
		}

		// The sub-context binds the type parameter to the original key, not
		// the remapped one.
		propertyType := p.childNodeParser.CreateType(mapped.Type, p.createSubContext(node, key, context), nil)

		newType := types.DerefAnnotatedType(propertyType)
		hasUndefined := false
		if union, ok := newType.(*types.UnionType); ok {
			numRemoved, unionWithoutUndefined := types.RemoveUndefined(union)
			hasUndefined = numRemoved > 0
			newType = unionWithoutUndefined
		}

		result = append(result, types.NewObjectProperty(
			jsValueToString(mappedKey.Value),
			types.PreserveAnnotation(propertyType, newType),
			p.isMappedPropertyRequired(node, hasUndefined),
		))
	}

	return result
}

func (p *MappedTypeNodeParser) getValues(node *ast.Node, keyListType *types.EnumType, context *Context) []*types.ObjectProperty {
	mapped := node.AsMappedTypeNode()
	var result []*types.ObjectProperty
	for _, value := range keyListType.Values {
		if value == nil {
			continue
		}
		valueType := p.childNodeParser.CreateType(
			mapped.Type,
			p.createSubContext(node, &types.LiteralType{Value: value}, context),
			nil,
		)
		result = append(result, types.NewObjectProperty(jsValueToString(value), valueType, p.isMappedPropertyRequired(node, false)))
	}
	return result
}

func (p *MappedTypeNodeParser) getAdditionalProperties(node *ast.Node, keyListType *types.UnionType, context *Context) any {
	if types.IsDeepLiteralUnion(keyListType) {
		return p.additionalProperties
	}

	var key types.Type
	for _, t := range keyListType.Types() {
		if _, isLiteral := types.DerefType(t).(*types.LiteralType); !isLiteral {
			key = t
			break
		}
	}

	if key != nil {
		if t := p.childNodeParser.CreateType(node.AsMappedTypeNode().Type, p.createSubContext(node, key, context), nil); t != nil {
			return t
		}
		return p.additionalProperties
	}

	return p.additionalProperties
}

func (p *MappedTypeNodeParser) createSubContext(node *ast.Node, key types.Type, parentContext *Context) *Context {
	subContext := NewContext(node)

	for _, parentParameter := range parentContext.Parameters() {
		subContext.PushParameter(parentParameter)
		subContext.PushArgument(parentContext.GetArgument(parentParameter))
	}

	subContext.PushParameter(node.AsMappedTypeNode().TypeParameter.Name().Text())
	subContext.PushArgument(key)

	return subContext
}

// jsValueToString mirrors JavaScript's String(value) for literal and enum
// values (string, number, boolean).
func jsValueToString(value any) string {
	switch v := value.(type) {
	case string:
		return v
	case bool:
		if v {
			return "true"
		}
		return "false"
	case float64:
		return types.NumberToString(v)
	case int:
		return types.NumberToString(float64(v))
	default:
		return fmt.Sprintf("%v", v)
	}
}
